src/main/java/org/generationcp/middleware/api/germplasm/pedigree/GermplasmPedigreeServiceImpl.java

```diff
@@ -0,0 +1,346 @@
+						// Use female parent to continue traversal if source is unknown
+					} else if (femaleGid != 0) {
+						node.setMaleParentNode(this.createUnknownParent());
+						this.addFemaleParentNode(node, level, germplasm.getFemaleParent(), false);
+					}
+				}
+			} else if (germplasm.getGnpgs() >= 2) {
+				// Get and add female and male parents for generative germplasm
+				this.addNodeForParents(node, level, germplasm, excludeDerivativeLines);
+			}
+		}
+		return node;
+	}
+
+	private void addNodeForParents(final GermplasmTreeNode node, final int level, final Germplasm germplasm,
+		final boolean excludeDerivativeLines) {
+		// For generative germplasm, do not add any node if both parents are UNKNOWN (GID=0)
+		if (!(germplasm.getGpid1() == 0 && germplasm.getGpid2() == 0)) {
+			this.addFemaleParentNode(node, level, germplasm.getFemaleParent(), excludeDerivativeLines);
+			if (germplasm.getGpid2() == 0) {
+				node.setMaleParentNode(this.createUnknownParent());
+			} else {
+				this.addMaleParentNode(node, level, germplasm.getMaleParent(), excludeDerivativeLines);
+			}
+
+			// If there are more parents, get and add each of them
+			if (germplasm.getGnpgs() > 2) {
+				final List<Germplasm> otherParents =
+					this.germplasmService.getProgenitorsWithPreferredName(germplasm.getGid());
+				for (final Germplasm otherParent : otherParents) {
+					final GermplasmTreeNode maleParentNode = new GermplasmTreeNode(otherParent);
+					node.getOtherProgenitors().add(this.addParents(maleParentNode, level - 1, otherParent, excludeDerivativeLines));
+				}
+			}
+		}
+	}
+
+	private void addNodeForKnownParents(final GermplasmTreeNode node, final int level, final Germplasm germplasm,
+		final boolean excludeDerivativeLines) {
+		this.addFemaleParentNode(node, level, germplasm.getFemaleParent(), excludeDerivativeLines);
+		this.addMaleParentNode(node, level, germplasm.getMaleParent(), excludeDerivativeLines);
+	}
+
+	private void addMaleParentNode(final GermplasmTreeNode node, final int level, final Germplasm germplasm,
+		final boolean excludeDerivativeLines) {
+		if (germplasm != null) {
+			final GermplasmTreeNode maleParentNode = new GermplasmTreeNode(germplasm);
+			node.setMaleParentNode(maleParentNode);
+			this.addParents(maleParentNode, level - 1, germplasm, excludeDerivativeLines);
+		}
+	}
+
+	private void addFemaleParentNode(final GermplasmTreeNode node, final int level, final Germplasm germplasm,
+		final boolean excludeDerivativeLines) {
+		if (germplasm != null) {
+			final GermplasmTreeNode femaleParentNode = new GermplasmTreeNode(germplasm);
+			node.setFemaleParentNode(femaleParentNode);
+			this.addParents(femaleParentNode, level - 1, germplasm, excludeDerivativeLines);
+		}
+	}
+
+	private GermplasmTreeNode createUnknownParent() {
+		return new GermplasmTreeNode(0, Name.UNKNOWN, null, null, null);
+	}
+
+	public Integer getNumberOfGenerations(final Germplasm germplasm, final Boolean includeDerivativeLine) {
```

Comment:
Can use primitive boolean for `includeDerivativeLine`

Reply:
Done